File: fragments/solana_rpc/solana_rpc_utils.py
import asyncio
import time
from solana.rpc.async_api import AsyncClient
from solana.rpc.commitment import Commitment
from fragments.env_vars import get_env_var


def init_rpc_client() -> AsyncClient:
    url = "http://127.0.0.1:8899"
    host = get_env_var("SOLANA_HOST")

    if host is not None:
        url = url.replace("127.0.0.1", host)

    client = AsyncClient(url, Commitment("confirmed"))

    return client


async def wait_for_slot(slot: int, timeout: float = 5.0) -> bool:
    client = init_rpc_client()
    deadline = time.time() + timeout

    while time.time() < deadline:
        current_slot = (await client.get_slot()).value
        if current_slot >= slot:
            return True
        await asyncio.sleep(0.2)

    return False

File: fragments/solana_rpc/solana_rpc_utils.go
package solana_rpc

import (
	"context"
	"ff/env_vars"
	"fmt"
	"log"
	"strings"
	"time"

	"github.com/gagliardetto/solana-go/rpc"
	"github.com/gagliardetto/solana-go/rpc/ws"
)

func getRpcUrl() string {
	localhost := "http://127.0.0.1:8899"

	if len(env_vars.GetEnvVar("CI")) == 0 {
		return localhost
	}

	return strings.Replace(localhost, "127.0.0.1", "solana", 1)
}

func getRpcSubscriptionsUrl() string {
	localhost := "ws://127.0.0.1:8900"

	if len(env_vars.GetEnvVar("CI")) == 0 {
		return localhost
	}

	return strings.Replace(localhost, "127.0.0.1", "solana", 1)
}

func InitRpcClient() *rpc.Client {
	url := getRpcUrl()
	client := rpc.New(url)

	return client
}

func InitRpcSubscriptionsClient() *ws.Client {
	url := getRpcSubscriptionsUrl()
	client, err := ws.Connect(context.Background(), url)

	if err != nil {
		log.Fatalf("Failed to connect to WebSocket: %v", err)
	}

	return client
}

func WaitForSlot(slot uint64, timeout *uint64) (bool, error) {
	client := InitRpcClient()
	to := uint64(5000)
	if timeout != nil {
		to = *timeout
	}
	deadline := time.Now().Add(time.Duration(to) * time.Millisecond)

	for {
		currentSlot, err := client.GetSlot(context.Background(), rpc.CommitmentConfirmed)
		if err != nil {
			return false, fmt.Errorf("GetSlot failed %v", err)
		}
		if currentSlot >= slot {
			return true, nil
		}
		if time.Now().After(deadline) {
			return false, nil
		}

		time.Sleep(200 * time.Millisecond)
	}
}
